File: Generator/ssh-askpass.py
# -*- coding: utf-8 -*-
# originally from https://raw.githubusercontent.com/theseal/homebrew-ssh-askpass/master/Formula/ssh-askpass.rb

import os
import plistlib
import re

import requests

url = 'https://raw.githubusercontent.com/theseal/homebrew-ssh-askpass/master/Formula/ssh-askpass.rb'
page = requests.get(url).text

caveats_flag = False
plist_flag = False

caveats_text = list()
plist_text = list()
for line in page.splitlines():
    if 'def caveats; <<~EOF' in line:
        caveats_flag = True
        continue
    if 'def plist; <<~EOS' in line:
        plist_flag = True
        continue

    if caveats_flag:
        if 'EOF' in line:
            caveats_flag = False
            continue
        caveats_text.append(line.strip())

    if plist_flag:
        if 'EOS' in line:
            plist_flag = False
            continue
        plist_text.append(line.strip())

CAVEATS = f'{os.linesep}      '.join(caveats_text).strip()
PLIST = f'{os.linesep}      '.join(plistlib.dumps(plistlib.loads(''.join(plist_text).strip().encode())).decode().strip().splitlines())  # pylint: disable=line-too-long

FORMULA = f'''\
class SshAskpass < Formula
  desc "SSH AskPass util for MacOS"
  homepage "https://github.com/theseal/ssh-askpass/"
  url "https://github.com/theseal/ssh-askpass/archive/v1.2.1.tar.gz"
  sha256 "285e52794db4d1e5d230b115db138cc9b5fcd5e0171c41e3b540e41c540bf357"

  def install
    bin.install "ssh-askpass"
  end

  def caveats
    text = <<~EOF
      {CAVEATS}
    EOF
    text
  end

  plist_options :startup => "true", :manual => "launch ssh-askpass"

  def plist
    file = <<~EOS
      {PLIST}
    EOS
    file
  end

  test do
    system "true"
  end
end
'''

with open('../Formula/ssh-askpass.rb', 'w') as file:
    file.write(FORMULA)

File: Formula/ssh-askpass.rb
class SshAskpass < Formula
  desc "macOS SSH AskPass utility"
  homepage "https://github.com/theseal/ssh-askpass/"
  url "https://github.com/theseal/ssh-askpass/archive/v1.5.0.tar.gz"
  sha256 "a519843354c3c0c20327dfacb918c8031ddba41d04245262cbbd0ae5524c341b"

  def install
    bin.install name.to_s
    # The label in the plist must be #{plist_name} in order for `brew services` to work
    # See https://github.com/Homebrew/homebrew-services/issues/376
    plist = "#{name}.plist"
    inreplace plist, %r{<string>com.github.theseal.ssh-askpass</string>}, "<string>#{plist_name}</string>"
    inreplace plist, %r{/usr/local/bin}, "#{opt_prefix}/bin"
    prefix.install plist => "#{plist_name}.plist"
  end

  def caveats
    <<~EOF
      NOTE: You will need to run the following to load the SSH_ASKPASS environment variable:
        brew services start #{name}
    EOF
  end

  test do
    system "true"
  end

  service do
    name macos: "#{plist_name}"
  end
end
